### PR TITLE
Support qwen1.5 in turbomind engine

### PR DESCRIPTION
**Pull request by lvhan028**
Note: window attention is not supported.

### CODE REVIEW

**Comment by zhyncs** [comment]
Hi @lvhan028 Does this pr support [Qwen1.5-0.5B](https://huggingface.co/Qwen/Qwen1.5-0.5B) and [Qwen1.5-1.8B](https://huggingface.co/Qwen/Qwen1.5-1.8B)? Thanks.

**Comment by lvhan028** [comment]
0.5B no, since its head_dim is 64 while the turbomind engine hardcode the head_dim 128
1.8B Yes.

**Comment by zhyncs** [comment]
> 0.5B no, since its head_dim is 64 while the turbomind engine hardcode the head_dim 128 1.8B Yes.

Thanks for your reply. Maybe we could also update the [supported table](https://github.com/InternLM/lmdeploy/blob/main/docs/en/supported_models/supported_models.md). Thanks.

**Comment by lvhan028** [comment]
> > 0.5B no, since its head_dim is 64 while the turbomind engine hardcode the head_dim 128 1.8B Yes.
> 
> Thanks for your reply. Maybe we could also update the [supported table](https://github.com/InternLM/lmdeploy/blob/main/docs/en/supported_models/supported_models.md). Thanks.

sure. updated.

File: docs/en/supported_models/supported_models.md [hunk 1]
```diff
@@ -2,37 +2,42 @@
...
-|     Code Llama     | 7B - 34B |    Yes    |   No    |  No   |
-|         YI         | 6B - 34B |    Yes    |   No    |  No   |
+|       Model        |    Size    | FP16/BF16 | KV INT8 | W4A16 |
+| :----------------: | :--------: | :-------: | :-----: | :---: |
+|       Llama        |  7B - 65B  |    Yes    |   Yes   |  Yes  |
+|       Llama2       |  7B - 70B  |    Yes    |   Yes   |  Yes  |
+|      InternLM      |  7B - 20B  |    Yes    |   Yes   |  Yes  |
+|     InternLM2      |  7B - 20B  |    Yes    |    -    |  Yes  |
+| InternLM-XComposer |     7B     |    Yes    |   Yes   |  Yes  |
+|        QWen        |  7B - 72B  |    Yes    |   Yes   |  Yes  |
```
Comment (zhyncs) [inline docs/en/supported_models/supported_models.md:12]: Shall we also update this? https://github.com/InternLM/lmdeploy/issues/1008

Comment (lvhan028) [reply docs/en/supported_models/supported_models.md:12]: OK

File: lmdeploy/turbomind/deploy/source_model/qwen.py [hunk 2]
```diff
@@ -112,3 +112,53 @@ def model_info(self):
...
+        result = []
+
+        for key in ['q', 'k', 'v']:
+            tensor = self.params.get(
+                f'model.layers.{i}.self_attn.{key}_proj.bias')
+            assert tensor is not None
+            result.append(tensor)
+
+        tensor = self.params.get(f'model.layers.{i}.self_attn.o_proj.weight')
+        dummy_oproj_bias = torch.zeros(tensor.shape[-1],
```
Comment (RunningLeon) [inline lmdeploy/turbomind/deploy/source_model/qwen.py:140]: maybe we could use `tensor.new_zeros(tensor.shape[-1])` to create the dumpy bias with same `dtype` and `device` as `tensor`

File: lmdeploy/turbomind/turbomind.py [hunk 3]
```diff
@@ -255,11 +256,15 @@ def _from_hf(self, model_source: ModelSource, model_path: str,
                 engine_config.model_format is None:
             engine_config.model_format = 'awq'
 
-        # when convert model, use architectures in config.json
-        model_arch = get_model_from_config(model_path)
+        assert is_supported(model_path), (
+            f'turbomind does not support {model_path}. '
+            f'Plz try pytorch engine instead.')
```
Comment (RunningLeon) [inline lmdeploy/turbomind/turbomind.py:261]: could remove `f` for this string

**Comment by RunningLeon** [comment]
opecompass results look OK

|模型| mmlu| gsm8k|
|:--|:--:| :--:|
|qwen1.5-7b-chat-HF |61.48| 55.65|
|qwen1.5-7b-chat-TB |61.47 |54.74|

**Review by RunningLeon — APPROVED** [review]
LGTM

**Comment by irexyc** [comment]
Have you tried `lmdeploy convert` command to convert the model?

**Comment by lvhan028** [comment]
> Have you tried `lmdeploy convert` command to convert the model?

not yet. I'll test it asap

**Comment by lvhan028** [comment]
```shell
lmdeploy convert qwen <qwen1.5-model-path>
```

**Comment by irexyc** [comment]
```
Traceback (most recent call last):
  File "/home/chenxin/miniconda3/envs/38/bin/lmdeploy", line 33, in <module>
    sys.exit(load_entry_point('lmdeploy', 'console_scripts', 'lmdeploy')())
  File "/home/chenxin/ws3/vl/lmdeploy/cli/entrypoint.py", line 26, in run
    args.run(args)
  File "/home/chenxin/ws3/vl/lmdeploy/cli/cli.py", line 151, in convert
    main(**kwargs)
  File "/home/chenxin/ws3/vl/lmdeploy/turbomind/deploy/converter.py", line 265, in main
    tokenizer_path = get_tokenizer_path(model_path, tokenizer_path)
  File "/home/chenxin/ws3/vl/lmdeploy/turbomind/deploy/converter.py", line 45, in get_tokenizer_path
    assert tokenizer_path, 'please supply tokenizer path by --tokenizer-path'
AssertionError: please supply tokenizer path by --tokenizer-path
```

**Comment by lvhan028** [comment]
> ```
> Traceback (most recent call last):
>   File "/home/chenxin/miniconda3/envs/38/bin/lmdeploy", line 33, in <module>
>     sys.exit(load_entry_point('lmdeploy', 'console_scripts', 'lmdeploy')())
>   File "/home/chenxin/ws3/vl/lmdeploy/cli/entrypoint.py", line 26, in run
>     args.run(args)
>   File "/home/chenxin/ws3/vl/lmdeploy/cli/cli.py", line 151, in convert
>     main(**kwargs)
>   File "/home/chenxin/ws3/vl/lmdeploy/turbomind/deploy/converter.py", line 265, in main
>     tokenizer_path = get_tokenizer_path(model_path, tokenizer_path)
>   File "/home/chenxin/ws3/vl/lmdeploy/turbomind/deploy/converter.py", line 45, in get_tokenizer_path
>     assert tokenizer_path, 'please supply tokenizer path by --tokenizer-path'
> AssertionError: please supply tokenizer path by --tokenizer-path
> ```

What's the command?

**Review by irexyc — APPROVED** [review]

**Comment by xiaoxiaoyuwen** [comment]
dose this support qwen1.5-14b-chat int8 kv cache ?
I run the calibrate cause error
> RuntimeError: Currently, quantification and calibration of Qwen2ForCausalLM are not supported.

**Comment by lvhan028** [comment]
hi, @xiaoxiaoyuwen 
we are going to remove offline kv int8 and adopt online kv int8 in PR #1377 
PR #1412 provides the guide. 
This feature will be released in the next version. Stay tuned.

**Pull request merged**